### PR TITLE
zsh: use libpcre2 instead of libpcre

### PR DESCRIPTION
**Pull request by oskarirauta**
Changes:
 - use libpcre2 instead of libpcre
 - patch Src/Modules/pcre.c: source from mainstream's git (23.9.2023)
 - patch configure.ac and config.h.in: replace libpcre with libpcre2 and remove missing AC_PROG_LN
 - patch Src/Makefile.in: use LN_S instead of LN
 - patch Test/V07pcre.ztst: add some libpcre2 specific tests
 - use autoconf: git version no longer supports traditional configure script, so it was easier to patch configure.ac

Maintainer: @msva
Compile tested: x86_64, latest git
Run tested: x86_64, latest git

Description:
zsh is my favourite shell. I found issue #22006 which mentioned zsh and that's where idea originated.
Patch can be removed when official zsh is next time updated (might take quite some time)

### CODE REVIEW

File: utils/zsh/Makefile [hunk 1]
```diff
@@ -20,6 +20,7 @@ PKG_LICENSE:=ZSH
 PKG_LICENSE_FILES:=LICENCE
 PKG_CPE_ID:=cpe:/a:zsh_project:zsh
 
+PKG_FIXUP:=autoreconf
```
Comment (BKPepe) [inline utils/zsh/Makefile:23]: Why this is required now? In the commit description, there is no such info on this one.

File: utils/zsh/patches/010-use-libpcre2.patch [hunk 2]
```diff
@@ -0,0 +1,638 @@
+--- a/Src/Makefile.in
```
Comment (BKPepe) [inline utils/zsh/patches/010-use-libpcre2.patch:1]: Each patch, which is getting added, usually in the ideal world should contain **Git header**.
Git header includes important details like: Commit author, Commit description, Commit subject, etc.

This helps to know who authored such changes and where it comes from. Is that backport, which will be removed with upcoming release.

File: utils/zsh/patches/010-use-libpcre2.patch [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
++++ b/Src/Makefile.in
+@@ -35,7 +35,7 @@ VPATH           = @srcdir@
+ sdir            = @srcdir@
+ sdir_top        = @top_srcdir@
+ INSTALL         = @INSTALL@
+-LN		= @LN@
++LN		= @LN_S@
```
Comment (BKPepe) [inline utils/zsh/patches/010-use-libpcre2.patch:8]: I wonder, why do we such change? This is not included in https://sourceforge.net/p/zsh/code/ci/b62e911341c8ec7446378b477c47da4256053dc0/ or if you rather prefer GitHub: https://github.com/zsh-users/zsh/commit/b62e911341c8ec7446378b477c47da4256053dc0

**Review by BKPepe — CHANGES_REQUESTED** [review]
In one your single commit, you are trying to split so many changes, which are not described anywhere and who knows why they are required and why do we need them.

Many things are really unrelated to use libpcre2 instead of pcre, which you tried to describe that in commit description. Please try to avoid that for the next time.

File: utils/zsh/Makefile (hunk 1 above)
Comment (oskarirauta) [reply utils/zsh/Makefile:23]: from commit:
> use autoconf: git version no longer supports traditional configure script, so it was easier to patch configure.ac

Provided patch is more or less a back port/cherry pick. As zsh's master no longer ships with configure script, instead uses auto-tools, there would had been a huge task of patching configure script that would also had been a huge task for reviewers to check what's changed. Zsh has supported auto-tools for quite some time.

from [docs](https://openwrt.org/docs/guide-developer/packages):

> Packages that are using auto-tools should work with simply “PKG_FIXUP:=autoreconf”.

> Why this is required now? In the commit description, there is no such info on this one.
This is required to use auto-tools over traditional configure script as said in documents.

File: utils/zsh/patches/010-use-libpcre2.patch (hunk 2 above)
Comment (oskarirauta) [reply utils/zsh/patches/010-use-libpcre2.patch:1]: So it should list myself as author, and description that says that Src/Makefile.in is patched to support openwrt which lacks AC_PROG_LN and just has AC_PROG_LN_S? I'd rather patch this to macros of auto-tools, so it would include AC_PROG_LN as well, but that would be one more PR; Even though bible says, that man should fill the country with his offsprings; it shouldn't be task for just one man.. If someone adds that to auto-tools defs, I'll gladly remove this change.

File: utils/zsh/patches/010-use-libpcre2.patch (hunk 3 above)
Comment (oskarirauta) [reply utils/zsh/patches/010-use-libpcre2.patch:8]: Yes, it is a work-around to be able to configure/build on openwrt build system which lacks AC_PROG_LN, which is then on @LN@ - but we do have LN_S which is enough, hard linking is not necessary.

**Comment by oskarirauta** [comment]
@BKPepe

There is reason for every change that this patch makes. When next release of zsh comes out, patch will be incompatible and all changes it makes, are merged to that release, except LN_S stuff, which is more a problem with openwrt build system. Hopefully though this patch will be then obsolete on that as well and AC_PROG_LN is included in auto-tools macros.

Nevertheless, with this you get what you requested for zsh; libpcre2 instead of libpcre. I am happy with that as well, as I use zsh personally, and libpcre in my case was depend only for zsh; now I no longer have to install it at all.

Please, clarify if you have proposals for changes. Most of what I wrote here/explained was already in the commit, but pretty shortly. Do you want me to expand descriptions of changes? What about patch or Makefile? Do you have proposals for changes? For the moment, I am not sure if you want or how you would want me to change anything.

**Comment by oskarirauta** [comment]
@BKPepe

Patch splitted to 2 patches. Commit message expanded.

**Comment by BKPepe** [comment]
Uhh, you wrote a lot. I need to go through all of it. I will not make it
today. Give me day or two. I am afraid that tomorrow my calendar is nearly
full and I am spending the day outside of the computer.

Dne ne 24. 9. 2023 19:42 uživatel Oskari Rauta ***@***.***>
napsal:

**Comment by oskarirauta** [comment]
@BKPepe 

Just answered to all the questions as explanatory as possible :)
I was kinda hoping for all requirements today... I'm off to work in few hours and nature of my work might require that I ain't back home until friday night...

I have possibility to work with my laptop through a vpn connection, but it's way better locally. Especially when I don't know if I have short days, or do they go past 16 hours.. If I am not answering briefly, then I just don't have time; bare patiently, and I'll get back on topic as soon as I am able to. But surely, if there's not much changes necessary, I probably can use my breaks to do some work; Only change that I think might be in place, is patch to that test file.. I just patched that as well, since it was relevant even when it's unnecessary for result.

**Comment by oskarirauta** [comment]
@msva

You should also review this..

**Comment by msva** [comment]
Well, I looked over the patch (unfortunately, I don't have enough spare time at the moment for full careful analisys).
And it looks kinda ok. At least, I didn't noticed any suspicious code.

But unfortunately, at the moment, I dont't even have a device to test it on (that patched version works fine).

I'll try to at least check that it builds correctly at next holidays.

**Comment by oskarirauta** [comment]
I am running with it. Builds and runs. Good luck.

**Comment by oskarirauta** [comment]
@BKPepe

Have you had a chance yet to check this?

File: utils/zsh/patches/010-use-libpcre2.patch [hunk 4]
```diff
@@ -0,0 +1,619 @@
+--- a/Src/Modules/pcre.c
```
Comment (BKPepe) [inline utils/zsh/patches/010-use-libpcre2.patch:1]: Hmm, this patch still requires some formalities. No one knows by looking into it that it is backport.

I know I could be annoying with this one, but it is in doc:
https://openwrt.org/submitting-patches#patch_checklist

and I really tired of patches, which do not have Git header. In main repo, there's plenty of them. Many of them hacks, which are really hard to be removed.

So, please you are not doing it only for me, but for others. It will be in Git history.

File: utils/zsh/patches/020-replace-missing-autotools-ln-with-ln_s.patch [hunk 5]
```diff
@@ -0,0 +1,21 @@
+--- a/Src/Makefile.in
```
Comment (BKPepe) [inline utils/zsh/patches/020-replace-missing-autotools-ln-with-ln_s.patch:1]: dtto, but this one in my point of view should be in separate commit

**Review by BKPepe — CHANGES_REQUESTED** [review]

File: utils/zsh/Makefile (hunk 1 above)
Comment (BKPepe) [reply utils/zsh/Makefile:23]: Ok, thanks! Would it be possible to keep this in separate commit to have it clear and simple? This from my point should not be in one huge patch, which would be really hard to track changes.

File: utils/zsh/patches/010-use-libpcre2.patch (hunk 3 above)
Comment (BKPepe) [reply utils/zsh/patches/010-use-libpcre2.patch]: Ok, in this case, it seems you are the author of this patch, thus please try to describe it in commit description in the patch itself (you can basically copy&paste what you said)

File: utils/zsh/patches/010-use-libpcre2.patch (hunk 2 above)
Comment (BKPepe) [reply utils/zsh/patches/010-use-libpcre2.patch]: Hmm, yes, could be. Sounds good

**Comment by BKPepe** [comment]
> There is reason for every change that this patch makes. When next release of zsh comes out, patch will be incompatible and all changes it makes, are merged to that release, except LN_S stuff, which is more a problem with openwrt build system. Hopefully though this patch will be then obsolete on that as well and AC_PROG_LN is included in auto-tools macros.

Yes, I bet that there is a reason for that, but it should be described in the patch itself that people will know, why it was required. Hiding it (not exactly, but you got the point) in multiple changes in one huge commit is not good idea.

> Nevertheless, with this you get what you requested for zsh; libpcre2 instead of libpcre. I am happy with that as well, as I use zsh personally, and libpcre in my case was depend only for zsh; now I no longer have to install it at all.

Yeah, I do not have installed libpcre anymore either!

> Please, clarify if you have proposals for changes. Most of what I wrote here/explained was already in the commit, but pretty shortly. Do you want me to expand descriptions of changes? What about patch or Makefile? Do you have proposals for changes? For the moment, I am not sure if you want or how you would want me to change anything.

One option is to move it from commit description to patch's commit description or have it on both places. 

About patches - because it is a really huge diff, I havent look at it so far. Quite busy with other things.

File: utils/zsh/Makefile (hunk 1 above)
Comment (BKPepe) [reply utils/zsh/Makefile:23]: But we are not using master branch and we are using still the same version. Dont you forget to add something?

File: utils/zsh/patches/010-use-libpcre2.patch (hunk 4 above)
Comment (oskarirauta) [reply utils/zsh/patches/010-use-libpcre2.patch:1]: There is no git header, since its source is not from git - only source code is from git. Surely as there is no official version with footprint of this source code, since it has code combined from both versions, this is a hack. Though when new version is released, removing this hack is as simple as removing patch.

Make a request change with requested change. For me this is completely self explanatory, but that's the point; I hardly ever read patch comments, instead I read the source code as C(++) is often simpler than descriptions that are nearly always vague for one who's mother language is not English.

But sure, make a change request with text of your desire.. And I'll apply it.

File: utils/zsh/patches/020-replace-missing-autotools-ln-with-ln_s.patch (hunk 5 above)
Comment (oskarirauta) [reply utils/zsh/patches/020-replace-missing-autotools-ln-with-ln_s.patch:1]: Then it should be in a commit before this, as this doesn't build without. And then this PR would become incompatible, as this is the patching order, that it's applied afterwards pcre2 patch.

Are you aware that byrocracy that you put me through is larger than making things work the way you wanted them to work? Just do concrete proposals on everything and I'll accept them all; after all, this is your project to update to completely to libpcre2 and I just volunteered with something like I've done on openwrt project for 20 years since my first wrt54gs gen1 which I still have 5 pieces in my attic... I do realise newer rules, but for someone like me it's a bit difficult since when patchwork was the way to go, it was already almost a reason to resign from contributing due to it becoming so difficult and trouble-some. Even though I talk (or write) a lot, I've always been more on the action and little less on the conversation side. There was a time when making things happen was the most trouble-some part and there was gratitude on what was given; these days I update software I maintain with 3 month cycles, since otherwise I'm writing changelogs and comments to every commit for all my free time, which more or less seems only purposeful slowing of things, because those changelogs and comments become obsolete on next update; just like this patch does. On next update, it is out dated and obsolete, won't even be possible to apply it, as it's already applied in mainstream code.

I'm currently working on another project(not openwrt related) and as tomorrow I need to go to school to update my licenses, I'm going to be very busy for this weekend; so, provide concrete examples please and I'll apply them. I talk a lot, but here's my weakest spot: documentation. Second weak spot is persuasion. Thank you.

**Comment by oskarirauta** [comment]
> > Please, clarify if you have proposals for changes. Most of what I wrote here/explained was already in the commit, but pretty shortly. Do you want me to expand descriptions of changes? What about patch or Makefile? Do you have proposals for changes? For the moment, I am not sure if you want or how you would want me to change anything.
> 
> One option is to move it from commit description to patch's commit description or have it on both places.

Yes, but make a concrete proposal with example that I can apply from here. That way you get it just the way you want, because you already know what this does and so on, so you know what you want there to read; instead of me guessing what that might be.

> About patches - because it is a really huge diff, I havent look at it so far. Quite busy with other things. Honestly, still curious why do we need to have them, because the version is same and what was backported is only libpcre2 support.🤔🤔

Depends where you look at it from; smaller than some of my private kernel patches....
But I'll help you here, take source and patch it. Then download current/latest git sources of zsh and compare with diff both versions of Src/Modules/pcre.c since it's the only actual source code file that has changed, other changes regard only compiling. If they haven't updated it further, you should see that patched file is identical to one on git source.
If you don't trust me, this will lift that burden and move it to zsh team.

I am also quite busy, and soon this branch get's old enough to have conflicts with repo and I absolutely don't have more time on my timeline for this. One solution would be to switch to current git source and only use my autotools patch, but I know what is the opinion on using trunk versions.

File: utils/zsh/Makefile (hunk 1 above)
Comment (oskarirauta) [reply utils/zsh/Makefile:23]: To succeed, we are swapping from traditional configure to configure.ac which makes build process different (equal to master branch minus some bells and whistles not available on release version). It's been like this in autotools version for quite some time, but we haven't utilised it until now.

I could apply another huge patch to fix traditional configure, that would be other way around this, but wouldn't that make things even more complicated? I attempted it in the first place and found out pretty soon that changes required in traditional configure script would be massive and definitely break if it ever gets updated due to amount of changes. So, I sticked to with current autotools version, traditional isn't even supported anymore; whole script has been removed from master branch. Probably been outdated long ago.

**Comment by oskarirauta** [comment]
@BKPepe

Patches updated with commentary and signed off.

**Comment by msva** [comment]
actually, patches aren't yet correct git-formatted patches, but it is no my business, if any, so it's on @BKPepe to review that part :-D

**Comment by oskarirauta** [comment]
@msva looks like it's a no go.

**Comment by Ansuel** [comment]
@oskarirauta you are welcome my guy 

https://gist.github.com/Ansuel/5f0901837fc65c01025f2c62002166a3

Also the second patch... should be reported in openwrt main repo as it seems ""easy"" to fix... @mpratt14 maybe can help? (autotools missing macro @LN@)

**Comment by mpratt14** [comment]
in the autotools patches called "relocatable" we end up hardcoding the paths anyway...

would it be better to just have LN_S exported from rules.mk?

**Comment by Ansuel** [comment]
@mpratt14 from what i can see the problem is that the aclocal.m4 is not getting loaded correctly... AC_PROG_LN is a non standard macro and is present in aclocal.m4 but is not getting evaluated???

**Comment by Ansuel** [comment]
@mpratt14 this is the zsh aclocal.m4 https://github.com/zsh-users/zsh/blob/master/aclocal.m4

**Comment by mpratt14** [comment]
maybe, I'm bad with the m4 files so I wouldn't know without really digging deep

I just happened to notice a coincidence that in rules.mk we define LN but not LN_S and it's also not exported...its a simple enough definition so I don't know a reason we shouldn't just export it

**Comment by mpratt14** [comment]
maybe, I'm bad with the m4 files so I wouldn't know without really digging deep

I just happened to notice a coincidence that in rules.mk we define LN but not LN_S and it's also not exported...its a simple enough definition so I don't know a reason we shouldn't just export it

**Comment by mpratt14** [comment]
just a guess, maybe the aclocal.m4 should be renamed because it's getting deleted before autoreconf

**Comment by Ansuel** [comment]
> just a guess, maybe the aclocal.m4 should be renamed because it's getting deleted before autoreconf

yes that is exactly the problem... aclocal is getting wiped... i see normally acinclude should be used... but renaming it doesn't fix it... now i'm checking makefile.in

**Comment by Ansuel** [comment]
I'm getting so triggered by these stupid thing....

**Comment by mpratt14** [comment]
add a hook to Hooks/Prepare/Post instead of patching that rename

maybe the makefile is listing aclocal.m4 in a list of local m4 files and that needs a patch?

**Comment by Ansuel** [comment]
@mpratt14 very fun indeed... for some reason renaming and patching the file is not liked by zsh build system... But i found the real cause...

Yes we were right from the start but it's not autoreconf that deleted the file... it's this bad boy https://github.com/openwrt/openwrt/blob/main/include/autotools.mk#L81 that by default remove aclocal.m4...

problem is that zsh used the default name to include custom macro and this is getting removed... so zsh fault and openwrt deleting files... 

by setting the variable empty everything compile correctly as the macro are correctly referenced...

**Comment by Ansuel** [comment]
@BKPepe took care of improving this... For me this is ready.
@oskarirauta thanks for creating the cherry-pick patch. Please check if you are ok with the change I added on top of yours.

**Comment by mpratt14** [comment]
@Ansuel I hate to be this picky, but there's a typo in your makefile comment

**Comment by oskarirauta** [comment]
I’m fine with all. Can’t review without problems now with Mobile device, but my work was pcre2 modification, autotools was just a necessary bad, and I’m glad you got it work out even better.Michael Pratt ***@***.***> kirjoitti 5.10.2023 kello 6.03:﻿
@Ansuel I hate to be this picky, but there's a typo in your makefile comment

**Comment by Ansuel** [comment]
@mpratt14 np but where?

**Comment by oskarirauta** [comment]
@Ansuel I finally got a chance to check. LGTM.

**Comment by mpratt14** [comment]
> @mpratt14 np but where? aclocak ?

yeah that

**Comment by oskarirauta** [comment]
@neheb

File: utils/zsh/patches/0001-HACK-add-temp-support-for-PCRE2.patch [hunk 6]
```diff
@@ -0,0 +1,648 @@
+From fabdb4164288b21f1323f783d257212e5aacecd4 Mon Sep 17 00:00:00 2001
```
Comment (BKPepe) [inline utils/zsh/patches/0001-HACK-add-temp-support-for-PCRE2.patch:1]: Regarding this patch, I think we can do it even more better.

I am not sure, why it is labeled as hack, when it seems by quick looking at this patch that it contains actually at least four commits from upstream, so we can do it as patch series as this is rather backport than hack.

Hack in my opinion is something, which is not included in upstream and won't be.

1. https://github.com/zsh-users/zsh/commit/b62e911341c8ec7446378b477c47da4256053dc0.patch
2. https://github.com/zsh-users/zsh/commit/f3f371deb376478176866fd770fbcf9bc0d0609f.patch
3. https://github.com/zsh-users/zsh/commit/b4d1c756f50909b4a13e5c8fe5f26f71e9d54f63.patch
4. https://github.com/zsh-users/zsh/commit/10bdbd8b5b0b43445aff23dcd412f25cf6aa328a.patch

Yeah, we need to remove the things, which we need, but still it will look better. :-)

Btw. Yes, I need to admit that this is a big improvement than it was, but it could be even better. 💪

**Review by BKPepe — CHANGES_REQUESTED** [review]

File: utils/zsh/patches/0001-HACK-add-temp-support-for-PCRE2.patch (hunk 6 above)
Comment (Ansuel) [reply utils/zsh/patches/0001-HACK-add-temp-support-for-PCRE2.patch]: @BKPepe indeed... that was my original idea and was blocked by the pkg fixup... Took care of backporting the patch and now it's much cleaner and direct... just 5 backport patch not touched (aside from the changelog commented out)

Comment (BKPepe) [reply utils/zsh/patches/0001-HACK-add-temp-support-for-PCRE2.patch]: See @oskarirauta you should learn from this guy. :) 👏 👏 Awesome work as always.

**Comment by BKPepe** [comment]
CI passed, thank you @Ansuel, let's give it time until evening, so someone can go through it. If no objections, otherwise I will merge this.

**Comment by mpratt14** [comment]
by the way, someone should inform the zsh developers that having a macro file in-project named "aclocal.m4" is deprecated and they should rename it or have an "m4" directory.

**Comment by Koston-0xDEADBEEF** [comment]
someone should tell openwrt people that zsh developers value backwards compatibility over changing file naming

**Comment by Ansuel** [comment]
bold to say this to the openwrt project itself where the base is to support old and deprecated devices...

**Comment by mpratt14** [comment]
one can have both backwards compatibility AND the new method, we don't have to choose one...

would a symlink named aclocal.m4 targeting a different named macro file work?
